src/coreclr/gc/gc.cpp

```diff
@@ -24922,221 +24943,210 @@ void gc_heap::check_heap_count ()
+        int step_up = (n_heaps + 1) / 2;
+        int extra_heaps = 1 + (n_max_heaps >= 32);
+        step_up = min (step_up, n_max_heaps - extra_heaps - n_heaps);
 
-            // on the way down, we essentially divide the heap count by 1.5
-            int step_down = (n_heaps + 1) / 3;
+        // on the way down, we essentially divide the heap count by 1.5
+        int step_down = (n_heaps + 1) / 3;
 
-            // estimate the potential time benefit of going up a step
-            float overhead_reduction_per_step_up = median_percent_overhead * step_up / (n_heaps + step_up);
+        // estimate the potential time benefit of going up a step
+        float overhead_reduction_per_step_up = median_percent_overhead * step_up / (n_heaps + step_up);
 
-            // estimate the potential time cost of going down a step
-            float overhead_increase_per_step_down = median_percent_overhead * step_down / (n_heaps - step_down);
+        // estimate the potential time cost of going down a step
+        float overhead_increase_per_step_down = median_percent_overhead * step_down / (n_heaps - step_down);
 
-            // estimate the potential space cost of going up a step
-            float space_cost_increase_per_step_up = percent_heap_space_cost_per_heap * step_up;
+        // estimate the potential space cost of going up a step
+        float space_cost_increase_per_step_up = percent_heap_space_cost_per_heap * step_up;
 
-            // estimate the potential space saving of going down a step
-            float space_cost_decrease_per_step_down = percent_heap_space_cost_per_heap * step_down;
+        // estimate the potential space saving of going down a step
+        float space_cost_decrease_per_step_down = percent_heap_space_cost_per_heap * step_down;
 
-    #ifdef STRESS_DYNAMIC_HEAP_COUNT
-            // quick hack for initial testing
-            int new_n_heaps = (int)gc_rand::get_rand (n_max_heaps - 1) + 1;
+#ifdef STRESS_DYNAMIC_HEAP_COUNT
+        // quick hack for initial testing
+        int new_n_heaps = (int)gc_rand::get_rand (n_max_heaps - 1) + 1;
 
-            // if we are adjusting down, make sure we adjust lower than the lowest uoh msl heap
-            if ((new_n_heaps < n_heaps) && (dynamic_heap_count_data.lowest_heap_with_msl_uoh != -1))
-            {
-                new_n_heaps = min (dynamic_heap_count_data.lowest_heap_with_msl_uoh, new_n_heaps);
-            }
-    #else //STRESS_DYNAMIC_HEAP_COUNT
-            int new_n_heaps = n_heaps;
-            if (median_percent_overhead > 5.0f)
+        // if we are adjusting down, make sure we adjust lower than the lowest uoh msl heap
+        if ((new_n_heaps < n_heaps) && (dynamic_heap_count_data.lowest_heap_with_msl_uoh != -1))
+        {
+            new_n_heaps = min (dynamic_heap_count_data.lowest_heap_with_msl_uoh, new_n_heaps);
+
+            // but not down to zero, obviously...
+            new_n_heaps = max (new_n_heaps, 1);
```

Comment:
```suggestion
            new_n_heaps = min ((dynamic_heap_count_data.lowest_heap_with_msl_uoh + 1), new_n_heaps);
```
since `dynamic_heap_count_data.lowest_heap_with_msl_uoh` is a heap index...

Reply:
Actually, no - the intention of this code is specifically to stress the situation where we are excluding a heap with a taken lock. But when the lowest heap is heap 0, we can't exclude that.

Reply:
ahh you are right, if it's heap#3, we'd want to exclude heap#3.